source/served/commands/format.d

```diff
@@ -131,3 +129,109 @@ string formatSnippet(string code, string[] dfmtArgs, SnippetLevel level = Snippe
 {
 	return backend.get!SnippetsComponent.format(code, dfmtArgs, level).getYield;
 }
+
+@protocolMethod("textDocument/rangeFormatting")
+TextEdit[] provideRangeFormatting(DocumentRangeFormattingParams params)
+{
+	import std.algorithm : filter;
+	import std.array : array;
+
+	return provideFormatting(DocumentFormattingParams(params.textDocument, params.options)).filter!(
+			(edit) => edit.range.isValidEditFor(params.range)
+	).array;
+}
+
+package bool isValidEditFor(const TextRange editRange, const TextRange formatRange)
+{
+	//dfmt off
+    return (editRange.start.line < formatRange.end.line
+        || (editRange.start.line == formatRange.end.line && editRange.start.character <= formatRange.end.character))
```

Comment:
add compare functions / operators for Position so this is easier to read

Reply:
I'm not sure how i should write this either xD

I think it's best to keep this simple, it's not needed elsewhere

Reply:
in the `struct Position` add:

```d
    int opCmp(const Position other) const
    {
        if (line < other.line)
            return -1;
        if (line > other.line)
            return 1;
        if (character < other.character)
            return -1;
        if (character > other.character)
            return 1;
        return 0;
    }
```

this allows you to compare positions like `a < b`

I believe it would be simpler to just have

```d
editRange.start <= formatRange.end && editRange.end >= formatRange.start
```

here.